### PR TITLE
feat(context): honor binding scope from @bind

### PR DESCRIPTION
**Pull request by raymondfeng**
The PR improves binding attributes as follows:

- Refine visibility for binding attributes
- Introduce `defaultScope` to only override binding scope if not set

Use case:

If a class is decorated with `@bind({scope: BindingScope.SINGLETON})`, we should honor it and not override it with a default scope. For example:

```
@bind({scope: BindingScope.SINGLETON})
class MyController {}

app.controller(MyController); // My controller should be bound as SINGLETON
```

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

**Comment by bajtos** [comment]
I am overloaded with PR reviews now, I'll try to find some time to review these changes later this week.

File: packages/context/src/__tests__/acceptance/bind-decorator.acceptance.ts [hunk 1]
```diff
@@ -72,4 +78,20 @@ describe('@bind - customize classes with binding attributes', () => {
       'controllers.my-controller',
     ]);
   });
+
+  it('honors binding scope from @bind', () => {
+    let binding = createBindingFromClass(MySingletonController, {
+      defaultScope: BindingScope.TRANSIENT,
```
Comment (jannyHou) [inline packages/context/src/__tests__/acceptance/bind-decorator.acceptance.ts]: nitpick: it would be good to mention the precedence in Binding's doc: https://loopback.io/doc/en/lb4/Binding.html#configure-binding-attributes-for-a-class

Comment (raymondfeng) [reply packages/context/src/__tests__/acceptance/bind-decorator.acceptance.ts]: Added.

File: packages/context/src/__tests__/acceptance/bind-decorator.acceptance.ts [hunk 2]
```diff
@@ -35,6 +35,12 @@ describe('@bind - customize classes with binding attributes', () => {
   })
   class MyController {}
 
+  @bind({
+    scope: BindingScope.SINGLETON,
+    tags: ['controller', {name: 'my-singleton-controller', type: 'controller'}],
+  })
+  class MySingletonController {}
```
Comment (bajtos) [inline packages/context/src/__tests__/acceptance/bind-decorator.acceptance.ts]: What is the minimal `@bind()` configuration needed to create a singleton controller? Would the following work?

```ts
@bind({scope: BindingScope.SINGLETON})
class MySingletonController {}
```

If it is intended to work, could you please add a test that will add such controller class via `app.controller` API and verify that name, type, namespace, etc. was set correctly?

File: packages/service-proxy/src/__tests__/unit/mixin/service.mixin.unit.ts [hunk 3]
```diff
@@ -78,7 +78,7 @@ describe('ServiceMixin', () => {
     const boundServices = myApp.find('services.*').map(b => b.key);
     expect(boundServices).to.containEql('services.GeocoderService');
     const binding = myApp.getBinding('services.GeocoderService');
-    expect(binding.scope).to.equal(BindingScope.TRANSIENT);
+    expect(binding.scope).to.be.undefined();
```
Comment (bajtos) [inline packages/service-proxy/src/__tests__/unit/mixin/service.mixin.unit.ts]: This looks like a backwards-incompatible change to me.

I think we should use a different approach:

- Keep the change setting the value of`binding._scope` to `undefined`
- Modify `binding.scope` getter to convert `undefined to `TRANSIENT`
- Add a new public method for applying the default scope, something along the following lines.

```ts
class Binding {
 // ..

  applyDefaultScope(scope: BindingScope) {
    if (this._scope) return;
    this._scope = scope;
  }
}
```

That will give use two Binding methods:
 - `binding.inScope` forces the scope the the value provided
 - `binding.applyDefaultScope` preserves explicitly configured scope and applies the given scope only when no explicit scope was defined

Comment (raymondfeng) [reply packages/service-proxy/src/__tests__/unit/mixin/service.mixin.unit.ts]: Good idea.

File: packages/core/src/application.ts [hunk 4]
```diff
@@ -50,6 +50,7 @@ export class Application extends Context {
       name,
       namespace: 'controllers',
       type: 'controller',
+      defaultScope: BindingScope.TRANSIENT,
```
Comment (bajtos) [inline packages/core/src/application.ts]: Do we have a test to verify this behavior?

Also since TRANSIENT is the default scope anyways, is it necessary to explicitly configure it here again? I am fine either way, just asking.

File: packages/core/src/application.ts [hunk 5]
```diff
@@ -80,7 +81,8 @@ export class Application extends Context {
       name,
       namespace: CoreBindings.SERVERS,
       type: 'server',
-    }).inScope(BindingScope.SINGLETON);
+      defaultScope: BindingScope.SINGLETON,
```
Comment (bajtos) [inline packages/core/src/application.ts]: Please add a test to verify this change.

I think this is the scenario:
 - A server  class is configured with TRANSIENT scope
 - The class is added to an app via `app.server` API
 - The server binding is expected to have TRANSIENT scope

File: packages/core/src/application.ts [hunk 6]
```diff
@@ -196,7 +198,8 @@ export class Application extends Context {
       name,
       namespace: 'components',
       type: 'component',
-    }).inScope(BindingScope.SINGLETON);
+      defaultScope: BindingScope.SINGLETON,
```
Comment (bajtos) [inline packages/core/src/application.ts]: Ditto - please add a test to verify this fix.

File: packages/repository/src/mixins/repository.mixin.ts [hunk 7]
```diff
@@ -71,6 +71,7 @@ export function RepositoryMixin<T extends Class<any>>(superClass: T) {
         name,
         namespace: 'repositories',
         type: 'repository',
+        defaultScope: BindingScope.TRANSIENT,
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts]: Do we have a test to verify this behavior?

Also since TRANSIENT is the default scope anyways, is it necessary to explicitly configure it here again? I am fine either way, just asking.

File: packages/repository/src/mixins/repository.mixin.ts [hunk 8]
```diff
@@ -122,7 +123,8 @@ export function RepositoryMixin<T extends Class<any>>(superClass: T) {
           name: name || dataSource.dataSourceName,
           namespace: 'datasources',
           type: 'datasource',
-        }).inScope(BindingScope.SINGLETON);
+          defaultScope: BindingScope.SINGLETON,
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts]: Please add a test to verify this fix.

File: packages/service-proxy/src/__tests__/unit/mixin/service.mixin.unit.ts [hunk 9]
```diff
@@ -3,7 +3,7 @@
 // This file is licensed under the MIT License.
 // License text available at https://opensource.org/licenses/MIT
 
-import {Application, Component, Provider, BindingScope} from '@loopback/core';
+import {Application, BindingScope, Component, Provider} from '@loopback/core';
```
Comment (bajtos) [inline packages/service-proxy/src/__tests__/unit/mixin/service.mixin.unit.ts]: Since there are no other changes in this file, I think it would be better to revert this reorder of imports.

File: packages/core/src/application.ts (hunk 4 above)
Comment (raymondfeng) [reply packages/core/src/application.ts]: I wanted to make it explicit.

File: packages/core/package.json [hunk 10]
```diff
@@ -12,7 +12,7 @@
     "clean": "lb-clean loopback-core*.tgz dist package api-docs",
     "pretest": "npm run build",
     "integration": "lb-mocha \"dist/__tests__/integration/**/*.js\"",
-    "test": "lb-mocha \"dist/__tests__/unit/**/*.js\" \"dist/__tests__/integration/**/*.js\" \"dist/__tests__/acceptance/**/*.js\"",
+    "test": "lb-mocha \"dist/__tests__/**/*.js\"",
```
Comment (bajtos) [inline packages/core/package.json]: This is unrelated to this pull request.

I don't have a strong opinion on how to specify test files. However, I would like us to be consistent in all packages, including our examples and CLI project templates.

Please revert this change here and open a new pull request to address the problem in a wider context.

File: packages/context/src/binding-inspector.ts [hunk 11]
```diff
@@ -191,29 +193,36 @@ export type BindingFromClassOptions = {
...
   typeNamespaceMapping?: TypeNamespaceMapping;
+  /**
+   * Default scope if the binding does not have an explicit scope
+   */
+  defaultScope?: BindingScope;
 };
 
 /**
  * Create a binding from a class with decorated metadata
  * @param cls A class
```
Comment (bajtos) [inline packages/context/src/binding-inspector.ts]: Ditto.

For example:

```suggestion
  * @param cls A class constructor (e.g. a Controller class) or a value provider (Provider class)
```

File: packages/core/package.json (hunk 10 above)
Comment (raymondfeng) [reply packages/core/package.json]: See https://github.com/strongloop/loopback-next/pull/2597
